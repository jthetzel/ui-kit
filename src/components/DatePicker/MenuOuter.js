import styled from 'react-emotion';

export default styled.div`
  position: absolute;
  left: 0;
  right: 0;
  top: 100%;
  z-index: 10;
  background: white;
  
  box-shadow: 0 10px 18px -2px rgba(0,0,0,0.75);
  text-align: center;
  
  .DayPickerKeyboardShortcuts_buttonReset {
    display: none;
  }
  
  .Range .DayPicker-Day--selected:not(.DayPicker-Day--start):not(.DayPicker-Day--end):not(.DayPicker-Day--outside) {
    background-color: #f0f8ff !important;
    color: #4a90e2;
  }
  .Range .DayPicker-Day {
    border-radius: 0 !important;
  }
`;
