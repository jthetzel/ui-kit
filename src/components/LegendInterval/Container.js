import styled from 'react-emotion';

export default styled.div`
  position: relative;
  display: flex;
  flex-wrap: wrap;
  margin: 0 20px;
`;
