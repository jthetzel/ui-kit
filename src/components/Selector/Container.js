import styled from 'react-emotion';

export default styled.div`
  position: relative;
  border-top: 1px solid black;
  border-bottom: 1px solid black;
  padding: 20px 0;
  user-select: none;
  
  &:hover {
    cursor: pointer;
  }
`;
