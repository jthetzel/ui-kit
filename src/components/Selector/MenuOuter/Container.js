import styled from 'react-emotion';

export default styled.ul`
  display: block;
  position: absolute;
  width: 200px;
  right: -220px;
  top: -17px;
  z-index: 10;
  background: white;
  padding: 10px 0;
  
  font-family: 'Noto Sans', serif;
  font-size: 16px;


  margin-left: 0;
  
  box-shadow: 0 10px 18px -2px rgba(0,0,0,0.75);
`;
