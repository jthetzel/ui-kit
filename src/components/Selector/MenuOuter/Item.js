import styled from 'react-emotion';

export default styled.li`
  display: flex;
  list-style-type: none;
  height: 50px;
  border-bottom: 1px solid gray;
  padding: 10px;
  box-sizing: border-box;
  align-items: center;
  user-select: none;
  
  background-color: ${({ isActive }) => (isActive ? '#f5f4f5' : '#FFFFFF')};

  &:hover {
    color: #223bf2;
  }

  &:last-child {
    border-bottom: 0;
  }
`;
