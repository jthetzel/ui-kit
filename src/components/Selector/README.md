Selector

```js
const options = [
  { value: 'one', label: 'One' },
  { value: 'two', label: 'Two' },
  { value: 'three', label: 'Three' },
  { value: 'four', label: 'Four' }
];

<div style={{ width: 300 }}>
  <Selector options={options} value={options[0].value} />
</div>
```