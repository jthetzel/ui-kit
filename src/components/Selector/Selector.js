import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import findIndex from '../../utils/findIndex';

import Container from './Container';
import MenuOuter from './MenuOuter/MenuOuter';

/**
 * @component
 */

class Selector extends PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      isOpen: this.props.isOpen,
      value: this.props.value
    };

    this.toggleMenu = this.toggleMenu.bind(this);
    this.handleOnClick = this.handleOnClick.bind(this);
    this.handleOutsideClick = this.handleOutsideClick.bind(this);
    this.setChildNodeRef = this.setChildNodeRef.bind(this);
  }

  componentWillMount() {
    const { options } = this.props;

    if (!options.length) {
      throw new Error('options is empty');
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.value !== this.props.value) {
      this.setState({ value: nextProps.value });
    }
  }

  toggleMenu() {
    if (!this.state.isOpen) {
      // attach/remove event handler
      document.addEventListener('click', this.handleOutsideClick, false);
    } else {
      document.removeEventListener('click', this.handleOutsideClick, false);
    }

    this.setState({ isOpen: !this.state.isOpen });
  }

  handleOnClick(value) {
    this.props.onChange(value);
    this.setState({ value });
  }

  handleOutsideClick(e) {
    const isDescendantOfRoot = this.childNode && this.childNode.contains(e.target);
    if (!isDescendantOfRoot) {
      this.toggleMenu();
    }
  }

  setChildNodeRef(ref) {
    this.childNode = ref;
  }

  render() {
    const { isOpen, value } = this.state;
    const { options } = this.props;

    const index = findIndex(options, o => o.value === value);
    const label = value && options[index].label;

    return (
      <Container
        innerRef={this.setChildNodeRef}
        onClick={this.toggleMenu}
      >
        {label || 'Выберите значение'}
        {
          isOpen &&
            <MenuOuter
              handleOnClick={this.handleOnClick}
              options={this.props.options}
              value={value}
            />
        }
      </Container>
    );
  }
}

Selector.propTypes = {
  options: PropTypes.arrayOf(PropTypes.shape({
    value: PropTypes.node.isRequired,
    label: PropTypes.node.isRequired
  })).isRequired,
  value: PropTypes.node,
  onChange: PropTypes.func,
  isOpen: PropTypes.bool
};

Selector.defaultProps = {
  value: null,
  isOpen: false,
  onChange: () => {}
};

export default Selector;
