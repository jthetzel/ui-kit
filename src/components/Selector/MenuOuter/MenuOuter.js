import React from 'react';
import Container from './Container';
import Item from './Item';

const MenuOuter = (props) => {
  const { value, onClick, options } = props;

  return (
    <Container onClick={e => e.stopPropagation()}>
      {
        options.map(option => (
          <Item
            key={option.value}
            style={{ background: value === option.value ? '#f5f4f5' : '#FFFFFF' }}
            onClick={onClick.bind(null, option.value)}
          >
            {option.label}
          </Item>
        ))
      }
    </Container>
  );
};

export default MenuOuter;
