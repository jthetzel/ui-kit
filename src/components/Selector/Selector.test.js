/* eslint function-paren-newline: 0 */

import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Selector from '../Selector';
import Container from '../Selector/Container';
import MenuOuter from '../Selector/MenuOuter/MenuOuter';
import MenuOuterContainer from '../Selector/MenuOuter/Container';
import Item from '../Selector/MenuOuter/Item';

const options = [
  { value: 'one', label: 'One' },
  { value: 'two', label: 'Two' },
  { value: 'three', label: 'Three' },
  { value: 'four', label: 'Four' }
];

const onChange = value => value;

test('Selector does not crash', () => {
  const tree = shallow(<Selector options={options} value={options[0].value} />);
  expect(toJson(tree)).toMatchSnapshot();
});

test('Selector with an empty array', () => {
  expect(() => shallow(<Selector options={[]} />).dive())
    .toThrowError('options is empty');
});

test('Selector test value', () => {
  const wrapper = mount(
    <Selector
      onChange={onChange}
      options={options}
      value={options[0].value}
    />);
  expect(wrapper.props().value).toEqual('one');
  wrapper.setProps({ value: 'two' });
  expect(wrapper.props().value).toEqual('two');
});

test('Selector simulate onClick', () => {
  const wrapper = shallow(
    <Selector
      onChange={onChange}
      options={options}
      value={options[0].value}
    />
  ).dive();
  const btn = wrapper.find(Container).first();
  expect(wrapper.state().isOpen).toEqual(false);
  btn.simulate('click');
  expect(wrapper.state().isOpen).toEqual(true);
});

test('Selector simulate change value', () => {
  const wrapper = shallow(
    <Selector
      onChange={onChange}
      options={options}
      value={options[0].value}
    />
  );
  expect(wrapper.props().value).toEqual(options[0].value);
  wrapper.setProps({ value: options[1].value });
  expect(wrapper.props().value).toEqual(options[1].value);
});

test('Selector simulate Item onClick', () => {
  const wrapper = shallow(
    <Selector
      onChange={onChange}
      options={options}
      value={options[0].value}
    />
  ).dive();
  wrapper.setState({ isOpen: true });
  const menuOuter = wrapper.find(MenuOuter).first();
  const itemLast = menuOuter.dive().find(Item).last();
  itemLast.simulate('click');
  expect(wrapper.state().value).toEqual(options[3].value);
});

test('Selector simulate click MenuOuter', () => {
  const wrapper = shallow(
    <Selector
      onChange={onChange}
      options={options}
      value={options[0].value}
    />
  ).dive();
  wrapper.setState({ isOpen: true });
  const menuOuter = wrapper.find(MenuOuter).first();
  const container = menuOuter.dive().find(MenuOuterContainer).first();
  container.simulate('click', { stopPropagation: () => {} });
  expect(wrapper.state().value).toEqual(options[0].value);
});
