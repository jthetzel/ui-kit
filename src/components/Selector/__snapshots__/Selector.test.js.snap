// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selector does not crash 1`] = `
<Selector
  isOpen={false}
  onChange={[Function]}
  options={
    Array [
      Object {
        "label": "One",
        "value": "one",
      },
      Object {
        "label": "Two",
        "value": "two",
      },
      Object {
        "label": "Three",
        "value": "three",
      },
      Object {
        "label": "Four",
        "value": "four",
      },
    ]
  }
  value="one"
/>
`;
